test-strategy.md

```diff
@@ -108,3 +108,83 @@ For instructions on how to write and run tests see the [contributing guide](http
 As we formalise our testing strategy and bring / keep our test up to standard, it is important that we have some principles on managing defects as they occur/ are reported. For now we can have one very simple rule:
 
 - Every bug fix should be accompanied by a test that failed before the fix and passes afterwards. This test should _typically_ be a dynamic code test but it could be a linting rule or new type if that is appropriate. There are always exceptions but we should think very carefully before ignoring this rule.
+
+## Playwright Tips
+
+Gradio uses [playwright](https://playwright.dev/docs/intro) to interact with gradio applications programmatically to ensure that both the frontend and backend function as expected.
+Playwright is very powerful but it can be a little intimidating if you haven't used it before.
+No one on the team is a testing expert so don't be afraid to ask if you're unsure how to do something.
+Likewise, if you learn something new about playwright, please share with the team! 
+
+### Tip 1 - Retrying Assertions
+
+Playwright tests are written imperitavely - first type into this textbox, then click this button, then check this textbox has this output.
+This is nice because it matches how users interact with Gradio applications.
+However, playwright carries out these steps much faster than any human can!
+This can cause you to check whether a textbox has the correct output before the server is finished processing the request.
+
+For this reason, playright ships with some [retrying assertions](https://playwright.dev/docs/test-assertions#auto-retrying-assertions).
+These assertions will retry until they pass or a timeout is reached, by default 5 seconds.
+So even if playwright checks a DOM element before the server is done, it gives the server a chance to finish by retrying.
+
+
+You can increase the timeout manually as well:
+
+```js
+// 5 seconds
+await expect(page.getByText('Hidden text')).toBeAttached({timeout?: 5000});
```

Comment:
For clarity, could you include what this test would look like without the retrying assertion?